lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,279 @@
+    }
+  }
+  return retArr;
+}
+
+// Get list of property from Function.prototype
+// and push to blacklistFunctionAttr;
+function getFunctionBlacklist(og) {
+  blacklistFunctionAttr = [];
+  FunctionId = og.lookup('Function', rootId);
+  functionId = og.lookup('prototype', FunctionId);
+  var functionProp = og.getObjectKeys(functionId);
+  for (var i = 0; i < functionProp.length; i += 1) {
+    blacklistFunctionAttr.push(cleanUp(functionProp[i]));
+  }
+}
+
+// Get list of property from Object.prototype
+// and push to blacklistObjectAttr;
+function getObjectBlacklist(og) {
+  blacklistObjectAttr = [];
+  ObjectId = og.lookup('Object', rootId);
+  objectId = og.lookup('prototype', ObjectId);
+  var objectProp = og.getObjectKeys(objectId);
+  for (var i = 0; i < objectProp.length; i += 1) {
+    blacklistObjectAttr.push(cleanUp(objectProp[i]));
+  }
+}
+
+// Search object id's prototype chain to find the
+// class it belongs to. The class is the prototype
+// with a contructor property.
+function getClassName(id, og) {
+  var ctorId = og.lookup('+constructor+', id);
+  if (ctorId) {
+    return og.getFunctionName(ctorId);
+  }
+  return og.getFunctionName(og.getPrototype(id));
+}
+
+// Remove blacklist properties and add Function and Object's APIs.
+function postProcess(apiCatalogs, og) {
+  for (let i = 0; i < blacklistProperties.length; i += 1) {
+    var prop = blacklistProperties[i];
+    if (apiCatalogs[prop]) {
+      delete apiCatalogs[prop];
+    }
+  }
+  // Add function and obejct blacklisted property
+  // to Function and Object interface.
+  if (!apiCatalogs[og.getFunctionName(FunctionId)]) {
+    apiCatalogs[og.getFunctionName(FunctionId)] = [];
+  }
+  apiCatalogs[og.getFunctionName(FunctionId)] =
+    apiCatalogs[og.getFunctionName(FunctionId)].concat(blacklistFunctionAttr);
+  if (!apiCatalogs[og.getFunctionName(ObjectId)]) {
+    apiCatalogs[og.getFunctionName(ObjectId)] = [];
+  }
+  apiCatalogs[og.getFunctionName(ObjectId)] =
+    apiCatalogs[og.getFunctionName(ObjectId)].concat(blacklistObjectAttr);
+}
+
+/** find API for interface of something with given id.
+  @param {ObjectGraph} og - The obejct graph.
+  @param {JSON} map - the json to store web api for interface
+  @param {int} id - the id currently visiting
+  @param {String} name - the property name from its parent node. (not necessarily interface name)
+  @param {Array} visited - a list of visted properties
+  @param {JSON} opt - know if this object is firstlevel or a __proto__ property.
+*/
+function extractAPI(og, map, id, name, visited, opt) {
+  opt = opt || {};
+  var firstLevel = opt.firstLevel || false;
+  var proto = opt.proto || false;
+
+  // Return if this object is already visited, except for first level objects.
+  // Still visit it if this is a firstLevel object, since
+  // two different firstlevel object can reference to one object.
+  // We don't want to miss any of them.
+  if (visited.indexOf(id) !== -1 && !firstLevel) return;
```

Comment:
Visited will get pretty big, no? Should we use a map of `{ <id> : <1 | true | something> }` instead?